jerry-core/ecma/builtin-objects/ecma-builtin-generator-prototype.c

```diff
@@ -194,52 +178,81 @@ ecma_builtin_generator_prototype_object_do (ecma_value_t this_arg, /**< this arg
- */
-static ecma_value_t
-ecma_builtin_generator_prototype_object_return (ecma_value_t this_arg, /**< this argument */
-                                                ecma_value_t return_arg) /**< return argument */
-{
-  return ecma_builtin_generator_prototype_object_do (this_arg, return_arg, ECMA_ITERATOR_RETURN);
-} /* ecma_builtin_generator_prototype_object_return */
+  vm_executable_object_t *executable_object_p = NULL;
 
-/**
- * The Generator.prototype object's 'throw' routine
- *
- * See also:
- *          ECMA-262 v6, 25.3.1.4
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_generator_prototype_object_throw (ecma_value_t this_arg, /**< this argument */
-                                                ecma_value_t throw_arg) /**< throw argument */
-{
-  return ecma_builtin_generator_prototype_object_do (this_arg, throw_arg, ECMA_ITERATOR_THROW);
-} /* ecma_builtin_generator_prototype_object_throw */
+  if (ecma_is_value_object (this_arg))
+  {
+    ecma_object_t *object_p = ecma_get_object_from_value (this_arg);
+
+    if (ecma_get_object_type (object_p) == ECMA_OBJECT_TYPE_CLASS)
+    {
+      ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
+
+      if (ext_object_p->u.class_prop.class_id == LIT_MAGIC_STRING_GENERATOR_UL)
+      {
+        executable_object_p = (vm_executable_object_t *) ext_object_p;
+      }
+    }
+  }
+
+  if (executable_object_p == NULL)
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a generator object."));
+  }
+
+  if (executable_object_p->extended_object.u.class_prop.extra_info & ECMA_EXECUTABLE_OBJECT_RUNNING)
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Generator is currently under execution."));
+  }
+
+  if (executable_object_p->extended_object.u.class_prop.extra_info & ECMA_EXECUTABLE_OBJECT_COMPLETED)
+  {
+    return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+  }
+
+  switch (builtin_routine_id)
+  {
+    case ECMA_GENERATOR_PROTOTYPE_ROUTINE_NEXT:
+    {
+      return ecma_builtin_generator_prototype_object_do (executable_object_p,
+                                                         arguments_list_p[0],
+                                                         ECMA_ITERATOR_NEXT);
+    }
+    case ECMA_GENERATOR_PROTOTYPE_ROUTINE_THROW:
+    {
+      return ecma_builtin_generator_prototype_object_do (executable_object_p,
+                                                         arguments_list_p[0],
+                                                         ECMA_ITERATOR_THROW);
+    }
+    case ECMA_GENERATOR_PROTOTYPE_ROUTINE_RETURN:
+    {
+      return ecma_builtin_generator_prototype_object_do (executable_object_p,
+                                                         arguments_list_p[0],
+                                                         ECMA_ITERATOR_RETURN);
```

Comment:
What about putting the type into a local variable and call `ecma_builtin_generator_prototype_object_do` once after the switch?